### PR TITLE
updated factory class

### PR DESCRIPTION
**Pull request by chillaq**
# .NET SDK

## Extra Notes

### CODE REVIEW

File: src/Splitio/Domain/RedisConfigurationValidator.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+using Splitio.Services.Shared.Classes;
+using System;
+
+namespace Splitio.Domain
+{
+    public static class RedisConfigurationValidator
+    {
+        private static readonly ISplitLogger _log = WrapperAdapter.Instance().GetLogger(typeof(RedisConfigurationValidator));
+
+        public static void ValidateRedisOptions(ConfigurationOptions options)
```
Comment (sanzmauro) [inline src/Splitio/Domain/RedisConfigurationValidator.cs:12]:
```suggestion
        public static void Validate(CacheAdapterConfigurationOptions config)
```

File: src/Splitio/Domain/RedisConfigurationValidator.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                throw new ArgumentNullException(nameof(options), "Redis Host and Port or Cluster Nodes should be set to initialize Split SDK in Redis Mode.");
+            }
+            if (options.CacheAdapterConfig.RedisClusterNodes != null)
+            {
+                if (options.CacheAdapterConfig.RedisClusterNodes.EndPoints.Count == 0)
+                {
+                    throw new ArgumentNullException(nameof(options), "Redis Cluster Nodes should have at least one host to initialize Split SDK in Redis Mode.");
+
+                }
+                if (options.CacheAdapterConfig.RedisClusterNodes.KeyHashTag == null)
```
Comment (sanzmauro) [inline src/Splitio/Domain/RedisConfigurationValidator.cs:25]:
```suggestion
                }

                if (string.IsNullOrEmpty(config.RedisClusterNodes.KeyHashTag))
```

File: src/Splitio/Domain/RedisConfigurationValidator.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                {
+                    options.CacheAdapterConfig.RedisClusterNodes.KeyHashTag = "{SPLITIO}";
+                    _log.Warn("Redis Cluster Hashtag is not set, will set its value to: {SPLITIO}.");
+                }
+                if (!string.IsNullOrEmpty(options.CacheAdapterConfig.Host))
```
Comment (sanzmauro) [inline src/Splitio/Domain/RedisConfigurationValidator.cs:30]:
```suggestion
                }

                if (!string.IsNullOrEmpty(options.CacheAdapterConfig.Host))
```

File: src/Splitio/Domain/RedisConfigurationValidator.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        private static readonly ISplitLogger _log = WrapperAdapter.Instance().GetLogger(typeof(RedisConfigurationValidator));
+
+        public static void ValidateRedisOptions(ConfigurationOptions options)
+        {
+            if ((string.IsNullOrEmpty(options.CacheAdapterConfig.Host) || string.IsNullOrEmpty(options.CacheAdapterConfig.Port)) && options.CacheAdapterConfig.RedisClusterNodes == null)
+            {
+                throw new ArgumentNullException(nameof(options), "Redis Host and Port or Cluster Nodes should be set to initialize Split SDK in Redis Mode.");
+            }
+            if (options.CacheAdapterConfig.RedisClusterNodes != null)
```
Comment (sanzmauro) [inline src/Splitio/Domain/RedisConfigurationValidator.cs:18]:
```suggestion
            }

            if (options.CacheAdapterConfig.RedisClusterNodes != null)
```

File: src/Splitio/Domain/RedisConfigurationValidator.cs (hunk 2 above)
Comment (sanzmauro) [inline src/Splitio/Domain/RedisConfigurationValidator.cs:25]:
```suggestion
                }

                if (options.CacheAdapterConfig.RedisClusterNodes.KeyHashTag == null)
```

File: src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs [hunk 5]
```diff
@@ -18,11 +19,13 @@ public void Validate(string apiKey)
             if (apiKey == string.Empty)
             {
                 _log.Error("factory instantiation: you passed and empty api_key, api_key must be a non-empty string.");
+                throw new ArgumentException("API Key must be set to initialize Split SDK.", nameof(apiKey));
```
Comment (sanzmauro) [inline src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs:22]:
```suggestion
                throw new Exception("API Key should be set to initialize Split SDK.");
```

File: src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs [hunk 6]
```diff
@@ -18,11 +19,13 @@ public void Validate(string apiKey)
...
             }
 
             if (apiKey == null)
             {
                 _log.Error("factory instantiation: you passed a null api_key, api_key must be a non-empty string.");
+                throw new ArgumentNullException(nameof(apiKey), "API Key must be set to initialize Split SDK.");
```
Comment (sanzmauro) [inline src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs:28]:
```suggestion
                throw new Exception("API Key should be set to initialize Split SDK.");
```

File: src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs (hunk 5 above)
Comment (sanzmauro) [reply src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs:22]: we can not change the Exception type.

File: src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs (hunk 6 above)
Comment (sanzmauro) [reply src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs:28]: we can not change the Exception type.

File: src/Splitio/Domain/RedisConfigurationValidator.cs (hunk 1 above)
Comment (sanzmauro) [inline src/Splitio/Domain/RedisConfigurationValidator.cs:12]: what happen if `CacheAdapterConfig` is null ?

**Review by sanzmauro — CHANGES_REQUESTED** [review]

File: src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs (hunk 5 above)
Comment (chillaq) [reply src/Splitio/Services/InputValidation/Classes/ApiKeyValidator.cs:22]: Sonar complained we are using too generic exception

File: src/Splitio/Domain/RedisConfigurationValidator.cs (hunk 1 above)
Comment (chillaq) [reply src/Splitio/Domain/RedisConfigurationValidator.cs]: We do check that in Factory before calling validate

File: Splitio.Redis/Services/Client/Classes/RedisClient.cs [hunk 7]
```diff
@@ -85,6 +76,10 @@ private void ReadConfig(ConfigurationOptions config)
             _config.FlagSetsInvalid = baseConfig.FlagSetsInvalid;
             _config.Mode = config.Mode;
             _config.FromCacheAdapterConfig(config.CacheAdapterConfig);
+            if (_config.RedisUserPrefix != null && _config.ClusterNodes != null && _config.ClusterNodes.KeyHashTag != null)
```
[the remark this answers is not in the document]
Comment (chillaq) [reply Splitio.Redis/Services/Client/Classes/RedisClient.cs:79]: ok

File: Splitio.Redis/Services/Domain/RedisConfig.cs [hunk 8]
```diff
@@ -39,6 +38,10 @@ public void FromCacheAdapterConfig(CacheAdapterConfigurationOptions options)
             TlsConfig = options.TlsConfig;
             PoolSize = options.PoolSize ?? 1;
             ClusterNodes = options.RedisClusterNodes;
+            if (RedisUserPrefix != null && ClusterNodes != null && ClusterNodes.KeyHashTag != null)
+            {
+                RedisUserPrefix = ClusterNodes.KeyHashTag + RedisUserPrefix;
+            }
```
Comment (sanzmauro) [inline Splitio.Redis/Services/Domain/RedisConfig.cs:44]:
```suggestion
            if (ClusterNodes != null)
            {
                RedisUserPrefix = ClusterNodes.KeyHashTag + RedisUserPrefix;
            }
```

Comment (sanzmauro) [reply Splitio.Redis/Services/Domain/RedisConfig.cs:44]: the `KeyHashTag` validation is already in `RedisConfigurationValidator` class.

File: src/Splitio/Domain/RedisConfigurationValidator.cs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            {
+                throw new Exception("Redis Host and Port or Cluster Nodes should be set to initialize Split SDK in Redis Mode.");
+            }
+
+            if (config.RedisClusterNodes != null)
+            {
+                if (config.RedisClusterNodes.EndPoints.Count == 0)
+                {
+                    throw new Exception("Redis Cluster Nodes should have at least one host to initialize Split SDK in Redis Mode.");
+
```
Comment (sanzmauro) [inline src/Splitio/Domain/RedisConfigurationValidator.cs:24]:
```suggestion
```

File: src/Splitio/Services/Client/Classes/SplitFactory.cs [hunk 10]
```diff
@@ -79,16 +76,11 @@ private void BuildSplitClient()
                     {
```
Comment (sanzmauro) [inline src/Splitio/Services/Client/Classes/SplitFactory.cs:76]: as we now create `RedisConfigurationValidator`, I suggest to move this validations: `if (_options.CacheAdapterConfig != null && _options.CacheAdapterConfig.Type == AdapterType.Redis)` into that class please, to have all validations in one place.

**Review by sanzmauro — CHANGES_REQUESTED** [review]

File: src/Splitio/Domain/RedisConfigurationValidator.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
```
Comment (sanzmauro) [inline src/Splitio/Domain/RedisConfigurationValidator.cs:29]: please remove empty line.

File: src/Splitio/Services/Client/Classes/SplitFactory.cs [hunk 12]
```diff
@@ -75,30 +72,24 @@ private void BuildSplitClient()
...
                     break;
                 case Mode.Consumer:
-                    if (_options.CacheAdapterConfig != null && _options.CacheAdapterConfig.Type == AdapterType.Redis)
-                    {
-                        try
-                        {
-                            if (string.IsNullOrEmpty(_options.CacheAdapterConfig.Host) || string.IsNullOrEmpty(_options.CacheAdapterConfig.Port))
-                            {
-                                throw new Exception("Redis Host and Port should be set to initialize Split SDK in Redis Mode.");
-                            }
```
Comment (sanzmauro) [inline src/Splitio/Services/Client/Classes/SplitFactory.cs:75]: please remove empty line.

**Review by sanzmauro — APPROVED** [review]
⚠️ Please remove empty lines before merge, I left the comments. ⚠️

File: src/Splitio/Domain/RedisConfigurationValidator.cs (hunk 11 above)
Comment (chillaq) [reply src/Splitio/Domain/RedisConfigurationValidator.cs:29]: ok

File: src/Splitio/Services/Client/Classes/SplitFactory.cs (hunk 12 above)
Comment (chillaq) [reply src/Splitio/Services/Client/Classes/SplitFactory.cs:75]: ok

**Pull request merged**